File: 2015/day1/day1.rs
// To run:
// rustc day1.rs | ./day1

use std::fs;

const INPUT_FILE : &str = "input.txt";

fn read_input() -> String {
    return fs::read_to_string(INPUT_FILE)
        .expect("Should have been able to read the file");
}

fn day1() {
    let input = read_input();
    
    let mut floor: i32 = 0;
    let mut position: i32 = 0;
    let mut basemend_found = false;
    for char in input.chars() {
        position += 1;
        if char == '(' {
            floor += 1;
        }
        if char == ')' {
            floor -= 1;
        }
        // Print position if basement
        if floor == -1 && !basemend_found {
            println!("Part 2 answer: {position}"); // 1771
            basemend_found = true;
        }
    }
    println!("Part 1 answer: {floor}"); // 138
    
}

fn main() {
    day1();
}

File: 2015/day1/day1.nim
# To run:
# nim r day1.nim

const input_file = "input.txt"

# Read input
import std/strutils

const input = staticRead(input_file).strip()

### Part 1

# Count the floors
var floor = 0

for c in input:
  if c == '(':
    floor += 1
  elif c == ')':
    floor -= 1

echo "Santa is on floor ", floor
# answer 138

### Part 2

floor = 0
for position, character in input:
    if character == '(':
        floor += 1
    if character == ')':
        floor -= 1
    
    # Check if basement
    if floor == -1:
        echo position+1 # Add +1 to position, since started from 0, not 1
        break
    
# answer 1771